plugins/Pandora_LayeredCalorimeterPlugin.cpp

```diff
@@ -0,0 +1,166 @@
+//
+// Pandora Layered Calorimeter Plugin
+//
+// The plugin builds LayeredCalorimeterData dynamically by
+// configuring detector's material properties, interaction lengths, and type classifications
+// ensuring proper adaptibility of a calorimeter with PandoraPFA irrespective of its geometry.
+//
+//==========================================================================
+*/
+
+#include <DD4hep/DetFactoryHelper.h>
+#include <DD4hep/Detector.h>
+#include <DD4hep/Handle.h>
+#include <DD4hep/Printout.h>
+#include <DD4hep/DetElement.h>
+#include <XML/Utilities.h>
+#include <DDRec/MaterialManager.h>
+#include <DD4hep/Factories.h>
+#include <DDRec/Vector3D.h>
+#include <DDRec/DetectorData.h>
+
+#include <string>
+#include <vector>
+#include <stdexcept>
+
+using namespace dd4hep;
+
+namespace {
+
+  static long addLayeredCalorimeterData(dd4hep::Detector& description, int argc, char** argv) {
+    const std::string LOG_SOURCE("LayeredCalorimeterPlugin");
+
+    // Ensure there are at least 4 arguments
+    if (argc < 7) {
+      printout(PrintLevel::ERROR, LOG_SOURCE, "Insufficient arguments. Expected: path, Rmin, Rmax, half_length,angle, and layer heights.");
+      return 1;
+    }
+
+    // Parse basic parameters
+    std::string subDetectorName = argv[0];
+    std::string type_flags = argv[1];
+    double Rmin = dd4hep::_toDouble(argv[2]);
+    double Rmax = dd4hep::_toDouble(argv[3]);
+    double half_length = dd4hep::_toDouble(argv[4]);
+    double angle = dd4hep::_toDouble(argv[5]);
+    std::string name_of_the_material = argv[6];
+
+    // Parse layer heights (all remaining arguments)
+    std::vector<double> layerHeights;
+    for (int i = 7; i < argc; ++i) {
+      layerHeights.push_back(dd4hep::_toDouble(argv[i]));
+    }
+
+    dd4hep::printout(PrintLevel::DEBUG, LOG_SOURCE, "Parsed: path=%s, Rmin=%f, Rmax=%f, dR=%f, angle=%f,layers=%zu",
+                    subDetectorName.c_str(), Rmin, Rmax, half_length, angle,layerHeights.size());
+
+    // Retrieve the detector element
+    auto caloDetElem = description.detector(subDetectorName);
+    if (!caloDetElem.isValid()) {
+      dd4hep::printout(PrintLevel::ERROR, LOG_SOURCE, "Invalid DetElement path: %s", subDetectorName.c_str());
+      return 1;
+    }
+     
+    unsigned int typeFlag = dd4hep::_toInt(type_flags);
+    dd4hep::printout(PrintLevel::INFO, LOG_SOURCE,"+++ setDetectorTypeFlags for detector :%s set to 0x%x",
+		     subDetectorName.c_str(),
+		     typeFlag);
+
+    caloDetElem.setTypeFlag(typeFlag);
+
+
+    // Create LayeredCalorimeterData
+    auto caloData = new dd4hep::rec::LayeredCalorimeterData;
+    caloData->layoutType = dd4hep::rec::LayeredCalorimeterData::BarrelLayout;
+    caloData->extent[0] = Rmin;
+    caloData->extent[1] = Rmax;
+    caloData->extent[2] = 0.0; // Barrel-specific value
+    caloData->extent[3] = half_length;
+
+    caloDetElem.addExtension<dd4hep::rec::LayeredCalorimeterData>(caloData);
```

Comment:
could you move this line after the print out "Successfully added LayeredCalorimeterData", at the end of the function?

In addition, to be discussed... do you think it is acceptable to remove any data extension attached by the detector constructor itself? To do it:
```
    if( ! caloDetElem->extensions.empty()) {
      dd4hep::printout(PrintLevel::WARNING, LOG_SOURCE, "Removing data extensions from %s", subDetectorName.c_str());
      caloDetElem->extensions.clear();
    }
```
This also requires an additional header:
```
#include <DD4hep/detail/DetectorInterna.h>
```

This way, the plugin becomes more robust against potential cases where other data extension is hardcoded in the detector constructor

Reply:
Wouldn't this cause a memory leak?
```
      caloDetElem->extensions.clear();
```

Reply:
> ... do you think it is acceptable/advisable to remove any data extension attached 

Actually, it is not advisable to remove all extensions. It might make sense to remove the extension of the type we add here. It might also make sense to throw an error and let the user figure out which extension they want to keep. Maybe add a flag to replace or not an existing extension.

Reply:
We need to add a convenience function in dd4hep to make it easier to remove specific extensions. This issue brought up here will be addressed in a later PR then.